little_raft/src/replica.rs

```diff
@@ -223,16 +275,46 @@ where
     }
 
     fn broadcast_append_entry_request(&mut self) {
-        self.broadcast_message(|peer_id: ReplicaID| Message::AppendEntryRequest {
-            term: self.current_term,
-            from_id: self.id,
-            prev_log_index: self.next_index[&peer_id] - 1,
-            prev_log_term: self.log[self.next_index[&peer_id] - 1].term,
-            entries: self.get_entries_for_peer(peer_id),
-            commit_index: self.commit_index,
+        self.broadcast_message(|peer_id: ReplicaID| {
+            match self.get_term_at_index(self.next_index[&peer_id] - 1) {
+                Ok(term) => Message::AppendEntryRequest {
+                    from_id: self.id,
+                    term: self.current_term,
+                    prev_log_index: self.next_index[&peer_id] - 1,
+                    prev_log_term: term,
+                    entries: self.get_entries_for_peer(peer_id),
+                    commit_index: self.commit_index,
+                },
+                Err(ReplicaError::LogCompacted) => {
+                    let snapshot = self.snapshot.as_ref().unwrap();
+                    Message::InstallSnapshotRequest {
+                        from_id: self.id,
+                        term: self.current_term,
+                        last_included_index: snapshot.last_included_index,
+                        last_included_term: snapshot.last_included_term,
+                        offset: 0,
+                        data: snapshot.data.clone(),
+                        done: true,
+                    }
```

Comment:
It's not clear to me how to use this API with ChiselStore because Little Raft maintains a single snapshot and that's going to be a whole database (that can keep on growing in size). So whenever we compact the log on the leader, we have to send the _whole_ database over?

Reply:
First, for clarification, log compaction happens on every node independently, not just leaders. Leaders are only special in that they _might_ retransmit a snapshot to followers that are lagging too far behind.

Now to your question: if the whole database is the state, then it sounds like yes, you would have to transmit it. Conceptually, how else can you recover your state?